### PR TITLE
Replaced levelOfAssurance parameter with confidenceLevel

### PR DESCRIPTION
**Pull request by abdusahin**

### CODE REVIEW

**Comment by tonytw1** [comment]
+1

**Comment by jenshalm** [comment]
There are still open questions around whether we can continue to use default values (it might break some users) and the config should be validated like before to produce errors at startup time for illegal CL values and not at runtime.

Do not merge yet!

File: src/main/scala/uk/gov/hmrc/play/auth/controllers/AuthConfig.scala [hunk 1]
```diff
@@ -81,11 +58,14 @@ trait AuthParamsControllerConfig {
       account = config.getAs[String]("account"),
       agentRole = config.getAs[String]("agentRole"),
       delegatedAuthRule = config.getAs[String]("delegatedAuthRule"),
-      levelOfAssurance = config.getAs[LevelOfAssurance]("levelOfAssurance").getOrElse(globalLOA)
+      confidenceLevel = config.getAs[Int]("confidenceLevel").orElse(GlobalConfidenceLevel).getOrElse(throw NoConfidenceLevelError)
     )
```
Comment (jenshalm) [inline src/main/scala/uk/gov/hmrc/play/auth/controllers/AuthConfig.scala]: Stacktraces get generated when an Exception is instantiated, not when it is thrown. It's therefore better not to reuse existing instances as their stacktrace might be wrong.

File: README.md [hunk 2]
```diff
@@ -6,26 +6,26 @@ A library used for communicating with the auth service to authorise requests
 from microservices
 
 
-### Assurance Level Configuration
-Level of assurance for micro services can be configured in two ways. One option is to configure at the controllers level so that it's applied to all controllers of the micro service
-The other option is fine configuring at each controller level. When both global level and controller level is defined, the individual controller level assurance takes precedence
+### Confidence Level Configuration
+Confidence level for micro services can be configured in two ways. One option is to declare it at the controllers level so that it's applied to all controllers of the micro service
+The other option is specifying it for each controller. When the both global level and controller level are defined, the individual controller's confidence level takes precedence
```
Comment (jenshalm) [inline README.md]: When the both -> When both

File: src/test/scala/uk/gov/hmrc/play/auth/microservice/filters/AuthorisationFilterSpec.scala [hunk 3]
```diff
@@ -185,26 +185,24 @@ class AuthorisationFilterSpec extends WordSpecLike with MatchersResults with Sca
...
+          override def controllerConfigs: Config = ConfigFactory.parseMap(configMap ++ config)
         }
 
       override lazy val authConnector = testAuthConnector
-
-
     }
 
-    val authFilter = new TestAuthorisationFilter
+    def authFilter(config:  Map[String, String] = Map.empty) = new TestAuthorisationFilter(config)
```
Comment (jenshalm) [inline src/test/scala/uk/gov/hmrc/play/auth/microservice/filters/AuthorisationFilterSpec.scala]: It looks as if this new param is not used anywhere. Why has it been introduced?

File: src/test/scala/uk/gov/hmrc/play/auth/microservice/filters/AuthorisationFilterSpec.scala [hunk 4]
```diff
@@ -185,26 +185,24 @@ class AuthorisationFilterSpec extends WordSpecLike with MatchersResults with Sca
...
+          override def controllerConfigs: Config = ConfigFactory.parseMap(configMap ++ config)
         }
 
       override lazy val authConnector = testAuthConnector
-
-
     }
 
-    val authFilter = new TestAuthorisationFilter
+    def authFilter() = new TestAuthorisationFilter()
```
Comment (jenshalm) [inline src/test/scala/uk/gov/hmrc/play/auth/microservice/filters/AuthorisationFilterSpec.scala]: sorry for the nitpick, but this should remain a val, making many of the changes above (from `authFilter` to `authFilter()`) obsolete.

**Comment by jenshalm** [comment]
LGTM now.

**Pull request merged**